### PR TITLE
Adding agent internal telemetry

### PR DESCRIPTION
**Pull request by iglendd**
### What does this PR do?
Adds ability to send and query cross-org telemetry on a few internal Agent metrics. Specifically
* Sending a few internal and built-in checks.*, logs.*, transactions.*, dogstatsd.* metrics
* Disabled by default.
* Default schedule sent 30 seconds after agent start and every 15 minutes after that

### Additional Notes

* Agent check telemetry (telemetry.check) is activated when agent_telemetry.enabled, but resulting metrics will not be send automatically.
* We need to makes sure that agent memory usage will not grow and RSS remains stables (e.g. in cases if large number of metrics is collected and thrown away from Gather() method.

### Describe how to test/QA your changes
* Set agent_telemetry.enabled: true
* Restart agent
* Wait a few minutes minute
* Go to datadog org UI
* Select Log Explorer menu
* Paste in a search bar `source:agent-telemetry` - you should see a handful of messages with `Agent status` and `Agent metric` content, which contain JSON payloads.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by buraizu — DISMISSED** [review]
Thanks for the PR, I've left some minor updates to the release note

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+        - name: checks.warnings
+        - name: checks.metrics_samples
+        - name: checks.events
+        - name: logs.decoded
+        - name: logs.processed
+        - name: logs.sent
+        - name: logs.network_errors
+        - name: logs.dropped
+        - name: logs.sender_latency
+        - name: logs.destination_http_resp
```
Comment (vickenty) [inline comp/core/agenttelemetry/agenttelemetryimpl/config.go:214]: This metric has `url` tag that may leak sensitive info. Is that expected?

Since all these metrics are defined elsewhere, it would be easy to add tags with sensitive information by mistake. Would it make sense to allow-list all tags?

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+	Name     string             `yaml:"name"`
+	Metric   *AgentMetricConfig `yaml:"metric,omitempty"`
+	Status   *AgentStatusConfig `yaml:"status"`
+	Schedule *Schedule          `yaml:"schedule"`
+
+	// compiled
+	statusExtraBuilder []jBuilder
+	metricsMap         map[string]*MetricConfig
+	excludeZeroMetric  bool
+	excludeTags        map[string]any
```
Comment (vickenty) [inline comp/core/agenttelemetry/agenttelemetryimpl/config.go:42]: It seems we only assign `struct{}{}` here, would this work?
```suggestion
	excludeTags        map[string]struct{}
```

File: pkg/status/render/templates/agent-telemetry-basic.tmpl [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+  {{- with .RunnerStats }}
+  "checks": {
+    "running_checks": {{.RunningChecks}},
+    "runs": {{.Runs}}
+  },
+  {{- end }}
+  {{- with .DogStatsdStats -}}
+  {{- $alreadyIterated := 0 }}
```
Comment (vickenty) [inline pkg/status/render/templates/agent-telemetry-basic.tmpl:9]: Would it be possible to create this json payload with `encoding/json` instead?

Also, would it make sense to list specific data items to include, rather than doing it all in bulk? So that future modifications to dogstatsd or checks do not accidentally expose sensitive data (which otherwise is fine in agent status output or inventories).

File: pkg/config/setup/config.go [hunk 4]
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
 	// The histogram buckets use to track the time in nanoseconds it takes for a DogStatsD listeners to push data to the server
 	config.BindEnvAndSetDefault("telemetry.dogstatsd.listeners_channel_latency_buckets", []string{})
 
+	// Agent Telemetry
+	// Enable various telemetry on the internals of the Agent.
+	// This will not tirgger any billable custom metrics.
+	config.BindEnvAndSetDefault("agent_telemetry.enabled", false)
```
[the remark this answers is not in the document]
Comment (iglendd) [reply pkg/config/setup/config.go:1092]: I do not have strong preference. When I started spend some time thinking of reusing telemetry name space or package but in the end decided against it due to drastically different semantics and desire to avoid breaking existing code and behavior. It is also a different code hierarchy now. I do hope that at some point I would be able to to merge either agent_telemetry into telemetry package/component or other way around.

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go (hunk 2 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:42]: > It seems we only assign struct{}{} here, would this work?

No, actually. Because excluded tag can be specified with or without value. And if there is a value I change the node to another map. See https://github.com/DataDog/datadog-agent/pull/22771/files#diff-4d02442e857d5db8627ad2e0fc6f7cccc9dfff7e0765db5448878d3cfc817935R282-R284. Before `any` I did have `interface{}` type before.

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go (hunk 1 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:214]: >  This metric has url tag that may leak sensitive info. Is that expected?

Not sure about danger of url in terms of privacy but honestly I do not know what urls can be appeared there. I presumened it was Datadog domain url. 

> Since all these metrics are defined elsewhere, it would be easy to add tags with sensitive information by mistake. Would it make sense to allow-list all tags?
True, it did concerned me. It certainly has big hole when privacy may be inadvertently violated  by any developer without even realizing it. I will add facility to specify explicit tag management. I did not have time to do it before.

File: pkg/status/render/templates/agent-telemetry-basic.tmpl (hunk 3 above)
Comment (iglendd) [reply pkg/status/render/templates/agent-telemetry-basic.tmpl:9]: > Would it be possible to create this json payload with encoding/json instead?

Can you please clarify what do you mean exactly. Specifically what to add/change. The template is rendered as a []byte and later umarlshalle by encodiog/json (https://github.com/DataDog/datadog-agent/pull/22771/files#diff-1801173c4f68bdc061d17c1a22740783fd535108faafeae5c5d63e02b049312eR264)

> Also, would it make sense to list specific data items to include, rather than doing it all in bulk? So that future modifications to dogstatsd or checks do not accidentally expose sensitive data (which otherwise is fine in agent status output or inventories).

Yes, it would I make a necessary adjustments. Thank you.

Comment (iglendd) [reply pkg/status/render/templates/agent-telemetry-basic.tmpl:9]: Ah, my main goal was to avoid hardcoding transformation processing and simplify its management by changing mappings in the template. Especially if I need a different types of transformations. The amount of input/output, and frequency of its execution makes efficiency question moot IMHO but certainly help flexibility of transformation modification. And it appears that what templates are used across of agent as well and some of them to generate JSON from JSON, no?

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+func Module() fxutil.Module {
+	return fxutil.Component(
+		fx.Provide(newAtel))
+}
+
+type atel struct {
+	cfgComp    config.Component
+	logComp    log.Component
+	telComp    telemetry.Component
+	statusComp status.Component
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:46]: I'm confused about the need to use the status component as part of this new component.

Usually how things operate for the status component is that it aggregates information from other components via [FX value groups](https://uber-go.github.io/fx/value-groups/) and then when the customer calls `agent status` we display the information of each component.

It feels a bit weird how this component is using the status component. Ex [logs agent status provider](https://github.com/DataDog/datadog-agent/blob/2a41ad20c3f5051f3f0b9db2e0c23bda630a5bd0/comp/logs/agent/agent.go#L71). In the example the logs agent defines a status provider which would be register by the status component and later render. 

@iglendd I would like to know more about why you decided to use the status component in this way.

Also, just to be clear there is a lack of documentation on how to register status providers. I'm currently in the middle of the migration, so I will write a document explaining how to do it shortly.

**Review by GustavoCaso — CHANGES_REQUESTED** [review]
@iglendd I would love to have chat about the use of the status component to make sure we are levering the status component correctly.

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+	}
+
+	var atCfg Config
+
+	// Parse agent telemetry config
+	atCfgMap := cfg.GetStringMap("agent_telemetry")
+	if len(atCfgMap) > 0 {
+		// Reconvert to string and back to object.
+		// Config.UnmarshalKey() is better but it did not work in some cases
+		atCfgBytes, err := yaml.Marshal(atCfgMap)
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/config.go:498]: I'm not entire sure how are config parsing logic works 100%. Does the first pass to the config parsed nested configuration? Mostly asking to make sure we avoid doing any extra `yaml.Marshal`

File: comp/core/agenttelemetry/agenttelemetryimpl/utils.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/itchyny/gojq"
+	dto "github.com/prometheus/client_model/go"
+)
+
+type jBuilder struct {
+	jqSource    *gojq.Query
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/utils.go:24]: lThe idea of running JQ directly sounds good. It has a penalty on our side that we need to bring a 3rd party dependency into the mix. 

Is there any other way to achieve the same logic without adding support for JQ?

File: pkg/config/utils/telemetry.go [hunk 8]
```diff
@@ -32,5 +37,11 @@ func IsCheckTelemetryEnabled(checkName string, cfg pkgconfigmodel.Reader) bool {
 
 // IsTelemetryEnabled returns whether or not telemetry is enabled
 func IsTelemetryEnabled(cfg pkgconfigmodel.Reader) bool {
-	return cfg.IsSet("telemetry.enabled") && cfg.GetBool("telemetry.enabled")
+	return cfg.IsSet("telemetry.enabled") && cfg.GetBool("telemetry.enabled") ||
+		(IsAgentTelemetryEnabled(cfg))
+}
+
+// IsAgentTelemetryEnabled returns whether or not agent telemetry is enabled
+func IsAgentTelemetryEnabled(cfg pkgconfigmodel.Reader) bool {
```
Comment (GustavoCaso) [inline pkg/config/utils/telemetry.go:45]: Every time we make a public function on this package tends to be use in many places 😢 Ideally in the future we would have a single configuration component 🎉 

Since this function is only use here could we make it private or inline it?

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go (hunk 5 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:46]: My intent was extract parts of status JSON which is reach source of Agent Telemetry, I am using template and JQ for that. I am not providing or contributing to the status (yet).

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go (hunk 6 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:498]: I have hard time to extract more complex configuration hierarchy. One method work but it seems to have limitations. My double marshal/unmarshall is a work around for that since I do want to have unmarshal into dedicated structure. I did left a comment of that sort.

File: pkg/config/setup/config.go [hunk 9]
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
 	// The histogram buckets use to track the time in nanoseconds it takes for a DogStatsD listeners to push data to the server
 	config.BindEnvAndSetDefault("telemetry.dogstatsd.listeners_channel_latency_buckets", []string{})
 
+	// Agent Telemetry
```
Comment (olivielpeau) [inline pkg/config/setup/config.go:1089]: Please mention here that the feature is experimental and should not be enabled unless prompted by Datadog Support

File: pkg/config/setup/config.go [hunk 10]
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
...
+	// Enable various telemetry on the internals of the Agent.
+	// This will not tirgger any billable custom metrics.
```
Comment (olivielpeau) [inline pkg/config/setup/config.go:1091]: Let's remove this mention for now, can be re-introduced once the feature is not experimental anymore

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go (hunk 1 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:255]: I believe that I have addressed concerns in sub-sequent commit by adding to configuration needs to explicitly specify associated tags. If not tags are specified, all of them will be dropped (but timeseries aggregated accordingly)

File: pkg/status/render/templates/agent-telemetry-basic.tmpl (hunk 3 above)
Comment (iglendd) [reply pkg/status/render/templates/agent-telemetry-basic.tmpl]: I did not change that logic (and was not intended) but I have made a few improvements in last commit since (per discussion with @GustavoCaso). I have moved rendering and template itself into agenttelemetry component. And I have requesting now only sections are used during rendering and no more.

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go (hunk 5 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:53]: I have addressed some of the concerns. Please review the last commit which reflects our discussion.

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+	// Validate template (optional with default "basic". "none" is also supported)
+	if len(p.Status.Template) == 0 {
+		p.Status.Template = "basic"
+	} else if p.Status.Template != "basic" && p.Status.Template != "none" {
+		return fmt.Errorf("profile '%s' template attribute can have 'basic' or 'none' value but %s is provided", p.Name, p.Status.Template)
+	}
+
+	// Compile status extra
+	var err error
+	p.statusExtraBuilder, err = compileJBuilders(p.Status.Extra)
```
[the remark this answers is not in the document]
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:401]: To me it looks like it would muddy config.go file content a little bit but otherwise it is not principal.

File: comp/core/agenttelemetry/agenttelemetryimpl/utils.go (hunk 7 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/utils.go:24]: We already use this module (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/jsonquery/jsonquery.go#L13), which is further indirectly used in https://github.com/DataDog/datadog-agent/blob/main/pkg/compliance/resolver.go#L384 and https://github.com/DataDog/datadog-agent/blob/main/cmd/agent/common/autodiscovery.go#L319. Note to myself. Switch to our own wrapper - jsonquery

File: pkg/config/utils/telemetry.go (hunk 8 above)
Comment (iglendd) [reply pkg/config/utils/telemetry.go]: Right, made it private. Ideally we would like to join these two configurations in the future someway. @vickenty suggested to use a slightly different naming to indicate similar lineage (see his comment above)

File: pkg/config/setup/config.go (hunk 9 above)
Comment (iglendd) [reply pkg/config/setup/config.go]: Agree, fixed in the last commit

File: pkg/config/setup/config.go (hunk 10 above)
Comment (iglendd) [reply pkg/config/setup/config.go]: Agree, fixed in the last commit

**Comment by iglendd** [comment]
@GustavoCaso and @vickenty please re-review when you have a chance. I am not planning to merge before Privacy review is completed but I need to run tests on staging. I would like to rebase ideally soon but afraid it will screw up reviews and its positions.

File: comp/core/status/statusimpl/status_test.go [hunk 12]
```diff
@@ -927,7 +927,7 @@ Status render errors
 
 	for _, testCase := range testCases {
 		t.Run(testCase.name, func(t *testing.T) {
-			bytesResult, err := statusComponent.GetStatusBySection(testCase.section, testCase.format, false)
+			bytesResult, err := statusComponent.GetStatusBySection([]string{testCase.section}, testCase.format, false)
```
Comment (GustavoCaso) [inline comp/core/status/statusimpl/status_test.go:930]: Could we add a test case that uses multiple sections?
Another possible test is to make sure if the section do not exists or is empty dopes not cause any issue.

File: comp/core/status/component.go [hunk 13]
```diff
@@ -21,8 +21,8 @@ const CollectorSection string = "collector"
 type Component interface {
 	// Returns all the agent status information for the format type
 	GetStatus(format string, verbose bool, excludeSection ...string) ([]byte, error)
-	// Returns only the agent status for the especify section and format type
-	GetStatusBySection(section string, format string, verbose bool) ([]byte, error)
+	// Returns only the agent status for the specified sections and format type
+	GetStatusBySection(sections []string, format string, verbose bool) ([]byte, error)
```
Comment (GustavoCaso) [inline comp/core/status/component.go:25]: Should we rename the function to `GetStatusBySections` ?

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+// renderAgentStatus renders (transform) input status JSON object into output status using the template
+func (a *atel) renderAgentStatus(p *Profile, inputStatus map[string]interface{}, statusOutput map[string]interface{}) {
+	// Render template if needed
+	if p.Status.Template == "none" {
+		return
+	}
+
+	templateName := "agent-telemetry-" + p.Status.Template + ".tmpl"
+	var b = new(bytes.Buffer)
+	err := status.RenderText(templatesFS, templateName, b, inputStatus)
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:368]: 🎉

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+	// If no status is configured nothing to report
+	if p.Status == nil {
+		return
+	}
+
+	a.logComp.Debugf("Collect Agent Status telemetry for profile %s", p.Name)
+
+	// Current "agent-telemetry-basic.tmpl" uses only "runneStats" and "dogstatsdStats" JSON sections
+	// These JSON sections are populated via "collector" and "DogStatsD" status providers sections
+	minimumReqSections := []string{"collector", "DogStatsD"}
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:463]: nit: Since we might have more templates in the future with different required sections. We could use a map to store template and status sections information

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry_test.go [hunk 16]
```diff
@@ -0,0 +1,495 @@
...
+
+func (s statusMock) GetStatus(string, bool, ...string) ([]byte, error) {
+	return []byte{}, nil
+}
+func (s statusMock) GetStatusBySection([]string, string, bool) ([]byte, error) {
+	return []byte{}, nil
+}
+func newStatusMock() statusMock {
+	return statusMock{}
+}
```
Comment (GustavoCaso) [inline comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry_test.go:127]: I'll add it to my TODO list 😄

File: comp/core/agenttelemetry/agenttelemetryimpl/utils.go (hunk 7 above)
Comment (GustavoCaso) [reply comp/core/agenttelemetry/agenttelemetryimpl/utils.go:24]: Thanks for the context.

If I understand we use JQ to be able to modify the metrcis we emit juts by modifying the configuration, which would mean that we are able to change things faster rather than deploying the agent.

I still think that having this feature been use by customer (Still experimental) and being this dynamic might lead to unwanted consequences. Maybe a customer runs a complex JQ query (performance) Or creates an Invalid JQ  (shoot themself in the foot). Hard to migrate from using JQ (Once customers use it it might be difficult to remove t)

I'm raising this concerns, because we might reconsider shipping the initial implementation without JQ and see how adoption is and validate if we need this functionality. 

For now we could remove the `extra` feature:
```
extra:
        runner.workers.count: '.runnerStats.Workers.Count'
        runner.workers.utilization: '[.runnerStats.Workers | .. | objects | .Utilization] | add'
```

Extract those values manually for the `basic` template.

File: comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go (hunk 15 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:463]: True, but I want to see it first.

File: comp/core/status/component.go (hunk 13 above)
Comment (iglendd) [reply comp/core/status/component.go:25]: Sure, that is easy

Comment (iglendd) [reply comp/core/status/component.go]: see it in last commit

File: comp/core/status/statusimpl/status_test.go (hunk 12 above)
Comment (iglendd) [reply comp/core/status/statusimpl/status_test.go]: Done in last commit

**Comment by iglendd** [comment]
@GustavoCaso I have addressed most of your review comments (please check it out). @vickenty can you finish your review please. I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.

**Comment by vickenty** [comment]
> I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.

I think merging works a bit better than rebasing once review has started, so if possible and practical, please merge.

**Comment by iglendd** [comment]
> > I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.
> 
> I think merging works a bit better than rebasing once review has started, so if possible and practical, please merge.

Of course. I will do merging, thank you.

**Comment by GustavoCaso** [comment]
@iglendd I would like to draw some attention to my previous comment https://github.com/DataDog/datadog-agent/pull/22771#discussion_r1514260175

I just realized one thing about allowing arbitrary JQ functionality to run against the status JSON output. Once customers start to use that functionality we won't be able to modify the status JSON output without fear of breaking any telemetry `extra` custom metrics. 

I really want to surface that to see if we could consider my suggestion of shipping the first experimental version with support for `extra` 

I'm more than happy to chat about this via Slack, Zoom or whatever works best for you

**Comment by iglendd** [comment]
> @iglendd I would like to draw some attention to my previous comment [#22771 (comment)](https://github.com/DataDog/datadog-agent/pull/22771#discussion_r1514260175)
> 
> I just realized one thing about allowing arbitrary JQ functionality to run against the status JSON output. Once customers start to use that functionality we won't be able to modify the status JSON output without fear of breaking any telemetry `extra` custom metrics.
> 
> I really want to surface that to see if we could consider my suggestion of shipping the first experimental version with support for `extra`
> 
> I'm more than happy to chat about this via Slack, Zoom or whatever works best for you

Chatted OOTB

File: comp/core/agenttelemetry/agenttelemetryimpl/utils.go (hunk 7 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/utils.go:24]: discussed in OOTB

File: pkg/config/setup/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iglendd) [reply pkg/config/setup/config.go:1130]: I am closing for now. We can always revisit it later.

File: pkg/status/render/templates/agent-telemetry-basic.tmpl (hunk 3 above)
Comment (iglendd) [reply pkg/status/render/templates/agent-telemetry-basic.tmpl]: We can always discuss it further if needed as I did with @GustavoCaso. I am closing it for now, please notify me if it is not acceptable at this time.

File: comp/core/agenttelemetry/agenttelemetryimpl/config.go (hunk 6 above)
Comment (iglendd) [reply comp/core/agenttelemetry/agenttelemetryimpl/config.go:498]: Closing per OOTB discussion

File: comp/README.md [hunk 17]
```diff
@@ -69,6 +69,12 @@ Package collector defines the collector component.
 Package core implements the "core" bundle, providing services common to all
 agent flavors and binaries.
 
+### [comp/core/agenttelemetry](https://pkg.go.dev/github.com/DataDog/datadog-agent/comp/core/agenttelemetry)
+
+*Datadog Team*: agent-metrics-logs
+
+Package agenttelemetry implements a component to generate agent telemetry
```
Comment (buraizu) [inline comp/README.md:76]:
```suggestion
Package agenttelemetry implements a component to generate Agent telemetry
```

**Review by buraizu — APPROVED** [review]
Thanks, approving with a minor update requested

File: comp/metadata/host/hostimpl/utils/host_windows.go [hunk 18]
```diff
@@ -38,29 +39,12 @@ var (
...
-	Platform             string `json:"platform"`        // ex: ubuntu, linuxmint
-	PlatformFamily       string `json:"platformFamily"`  // ex: debian, rhel
-	PlatformVersion      string `json:"platformVersion"` // version of the complete OS
-	KernelVersion        string `json:"kernelVersion"`   // version of the OS kernel (if available)
-	KernelArch           string `json:"kernelArch"`
-	VirtualizationSystem string `json:"virtualizationSystem"`
-	VirtualizationRole   string `json:"virtualizationRole"` // guest or host
-	HostID               string `json:"hostid"`             // ex: uuid
-}
-
```
Comment (GustavoCaso) [inline comp/metadata/host/hostimpl/utils/host_windows.go:57]: @iglendd is this change intentional? 

I'm going to flag to @hush-hush who wrote the host implementation to make sure we are ok with this change

**Review by GustavoCaso — APPROVED** [review]
Len changes looks good. 

We have discussed extensible over Zoom and Slack. The use of the status component and JQ is to enable sending telemetry metrics as easy as possible. Knowing that feature is experimental and that is subject to change in the future, I would like for us to reconsider in the future the use of a similar pattern from the status component as each component registering there own telemetry metrics. Ensuring decoupling between components. Today the telemetry component has strong dependency on the status component. Adding new information for the telemetry agent means having to add it the status provider of a component. 
If we have that in mind and in future iterations we are able to remove that dependency I think we would be a in a better position to evolve the telemetry component in isolation. 

One more thing that would interesting to add in the PR description is amore detail explanation of the QA steps. ex. `Paste in a search bar source:agent-telemetry` Which search bar? Having as much detail information on how todo the QA the better. 

Len, thanks for the patience with all my comments and for taking the time to explain me the needs for the telemetry work.

File: comp/metadata/host/hostimpl/utils/host_windows.go (hunk 18 above)
Comment (hush-hush) [reply comp/metadata/host/hostimpl/utils/host_windows.go:57]: Looking through the history this was linked to no longer using WMI and removing usage of gopsutil from windows (#1725).

This code has been moved a number of times, we should definitely add a comment to explain this.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: comp/README.md (hunk 17 above)
Comment (iglendd) [reply comp/README.md]: Thank you @buraizu.  The change is accepted.

**Review by hush-hush — APPROVED** [review]

**Comment by iglendd** [comment]
> Knowing that feature is experimental and that is subject to change in the future, I would like for us to reconsider in the future the use of a similar pattern from the status component as each component registering there own telemetry metrics. Ensuring decoupling between components. Today the telemetry component has strong dependency on the status component. Adding new information for the telemetry agent means having to add it the status provider of a component. If we have that in mind and in future iterations we are able to remove that dependency I think we would be a in a better position to evolve the telemetry component in isolation.

Definitely. I also planning to start using our DefaultForwarder to replace portion of sender (cc @hush-hush) 

> One more thing that would interesting to add in the PR description is amore detail explanation of the QA steps. ex. `Paste in a search bar source:agent-telemetry` Which search bar? Having as much detail information on how todo the QA the better.

I have added a few more items in the step, I hope it is a bit better.
